### ISSUE TITLE
Underlay broken in pdf if turned on after sheet is created

### ISSUE THREAD

**Issue by sboddy** [IfcOpenShell/IfcOpenShell#5571]
Blender 4.2.2, Bonsai 0.8.1-alpha241011, Inkscape 1.3.2, OS: Ubuntu 22.04 LTS

1. Clear out any drawing or underlays from drawings and sheets folder
2. In prefs set the SVG->PDF command to `[["inkscape","svg","-o","pdf"]]`
3. Create an elevation without underlay and create the drawing (displays svg in default system viewer)
4. Create a sheet, add the elevation to it, and refresh it (displays linework on sheet in your default pdf viewer)
5. On the drawing, activate underlay, and create the drawing (displays svg in default system viewer, now with underlay)
6. Refresh the sheet.

Expected result:
![image](https://github.com/user-attachments/assets/4b4751eb-e73c-423c-a090-a167e39e14d5)

Actual result:
![image](https://github.com/user-attachments/assets/172e681e-b0ed-4552-97f4-3f30000a8b48)

When I loaded the svg into interactive Inkscape it warns of broken links, but that it has fixed them, and the underlays are present in Inkscape:
![image](https://github.com/user-attachments/assets/24818c2f-8ea5-47a9-b3c3-3b346e186511)

The problem appears to be that the underlay image does not get copied to the sheets folder after the drawings underlay is set to on in step 5. Nor does it get copied when the sheet is refreshed in step 6. On the command-line Inkscape is not searching and fixing these broken links, so you get the big red X's in the resulting pdf.

Workarounds are:
- ~~delete the svg sheets and drawings from disk, or~~ [Edit: just tried and didn't work]
- create another sheet using the drawing with underlay turned on, so that the underlay image is copied to the sheets folder. (~~Lingering question mark over whether subsequent changes to the drawings underlay would show up on the first sheet, or if it would only update when the second sheet is refreshed~~ [Edit: the drawing does not update when first sheet refreshed, only the second, so this is confusing too.])
- delete sheet in Bonsai and recreate it from scratch

Please excuse the crappy diagram. I'm still in the "figuring stuff out" phase. Making it pretty is a long way off.

**Comment by theoryshaw** [comment]
do any of the drawing names, have any characters like `(` or `&` in them?  I noticed there 'are' characters in the drawing name that break the link, sometimes.

**Comment by sboddy** [comment]
Nope. the elevation had a very original name (SOUTH ELEVATION) but all letters or a space. I didn't mention here, but I retested several times with new sheets with a single drawing and reproduced the issue easily following the steps above. Admittedly I didn't try it in a fresh file, but having just repeated the steps above in a fresh file, I still get the broken image in the resulting PDF. It does occur to me that the big red X is being done by the evince PDF viewer used under Ubuntu. I've no idea how Acrobat Reader displays it.
[A00 - UNTITLED.pdf](https://github.com/user-attachments/files/17372347/A00.-.UNTITLED.pdf)

**Comment by sboddy** [comment]
So this bugged me again today, so I took a bit deeeper look, knowing a little more about the code. I see two issues when activating an underlay _after_ adding it to a sheet.
1. The sheet does not get an IFCDOCUMENTREFERENCE to the RASTER image, i.e.
```
#137=IFCDOCUMENTINFORMATION('A00','UNTITLED',$,$,$,$,'SHEET',$,$,$,$,$,$,$,$,$,$);
#138=IFCRELASSOCIATESDOCUMENT('2ZAzfdsp53P8ErI4q5pRw5',$,$,$,(#1),#137);
#139=IFCDOCUMENTREFERENCE('layouts/A00 - UNTITLED.svg','X',$,'LAYOUT',#137);
#140=IFCDOCUMENTREFERENCE('layouts/titleblocks/A1.svg','X',$,'TITLEBLOCK',#137);
#141=IFCDOCUMENTREFERENCE('drawings/MY STOREY PLAN.svg','1',$,'DRAWING',#137);
#142=IFCDOCUMENTREFERENCE('sheets/A00 - UNTITLED.svg','X',$,'SHEET',#137);
#143=IFCDOCUMENTREFERENCE('sheets/MY STOREY PLAN-underlay.png','X',$,'RASTER',#137);
```
ID 143 here is when the underlay is activated when the sheet is created.
```
#127=IFCDOCUMENTINFORMATION('A00','UNTITLED',$,$,$,$,'SHEET',$,$,$,$,$,$,$,$,$,$);
#128=IFCRELASSOCIATESDOCUMENT('2YBmuGt$14gOhR1a23Cdwk',$,$,$,(#1),#127);
#129=IFCDOCUMENTREFERENCE('layouts/A00 - UNTITLED.svg','X',$,'LAYOUT',#127);
#130=IFCDOCUMENTREFERENCE('layouts/titleblocks/A1.svg','X',$,'TITLEBLOCK',#127);
#131=IFCDOCUMENTREFERENCE('drawings/MY STOREY PLAN.svg','1',$,'DRAWING',#127);
#132=IFCDOCUMENTREFERENCE('sheets/A00 - UNTITLED.svg','X',$,'SHEET',#127);
```
There is no similar RASTER reference when created without underlay (obviously!). However, when setting the underlay to true, no equivalent line is added. I can manually add one using the Project/Documents panel:
```
#168=IFCDOCUMENTREFERENCE('sheets/MY STOREY PLAN-underlay.png','X',$,'RASTER',#127);
```
but... it still doesn't work because of the next issue...

2. The layout gets created without the image(href/data-type) info for the raster image:
![image](https://github.com/user-attachments/assets/f301d5de-8a87-4499-8ebd-a6b5433a27a1)
It is this information that is used to find and copy the underlay from the drawing folder to the sheets folder here:
https://github.com/IfcOpenShell/IfcOpenShell/blob/1d271611c4caae44dac9ef434596b5aec40c12cf/src/bonsai/bonsai/bim/module/drawing/sheeter.py#L388-L392
So when the drawing is added to the layout without an underlay, the underlay will never work properly. The workaround (a bit annoying) is to remove the drawing from any sheets, and re-add and reposition them again.

I'm not sure if there is a sensible fix for this, or if simply having better feedback is the way to go, i.e. a pop-up warning that drawings will break on sheets that do not reference the RASTER when activating the underlay.

**Comment by theoryshaw** [comment]
I'm not sure, but wonder in general, if underlays should be embedded by default. 
https://community.osarch.org/discussion/comment/22577#Comment_22577

**Comment by sboddy** [comment]
Confirmed fixed for me.
